erpnext/regional/report/ksa_vat/ksa_vat.py

```diff
@@ -49,140 +49,183 @@ def get_columns():
-	total_tax = 0
+	tax_details = {}
+	for d in settings:
+		tax_details[d.account] = {
+			"title": d.title,
+			"taxable_amount": 0.00,
+			"adjustment_amount": 0.00,
+			"total_tax": 0.00
+		}
+
 	# Fetch All Invoices
-	invoices = frappe.get_all(doctype,
-	filters ={
-		'docstatus': 1,
-		'posting_date': ['between', [from_date, to_date]]
-	}, fields =['name', 'is_return'])
-
-	for invoice in invoices:
-		invoice_items = frappe.get_all(f'{doctype} Item',
-		filters ={
-			'docstatus': 1,
-			'parent': invoice.name,
-			'item_tax_template': vat_setting.item_tax_template
-		}, fields =['item_code', 'net_amount'])
-
-		for item in invoice_items:
-			# Summing up total taxable amount
-			if invoice.is_return == 0:
-				total_taxable_amount += item.net_amount
-
-			if invoice.is_return == 1:
-				total_taxable_adjustment_amount += item.net_amount
-
-			# Summing up total tax
-			total_tax += get_tax_amount(item.item_code, vat_setting.account, doctype, invoice.name)
-
-	return total_taxable_amount, total_taxable_adjustment_amount, total_tax
-
-
-
-def append_data(data, title, amount, adjustment_amount, vat_amount, company_currency):
-	"""Returns data with appended value."""
-	data.append({"title": _(title), "amount": amount, "adjustment_amount": adjustment_amount, "vat_amount": vat_amount,
-		"currency": company_currency})
-
-def get_tax_amount(item_code, account_head, doctype, parent):
-	if doctype == 'Sales Invoice':
-		tax_doctype = 'Sales Taxes and Charges'
-
-	elif doctype == 'Purchase Invoice':
-		tax_doctype = 'Purchase Taxes and Charges'
-
-	item_wise_tax_detail = frappe.get_value(tax_doctype, {
-		'docstatus': 1,
-		'parent': parent,
-		'account_head': account_head
-	}, 'item_wise_tax_detail')
-
-	tax_amount = 0
-	if item_wise_tax_detail and len(item_wise_tax_detail) > 0:
-		item_wise_tax_detail = json.loads(item_wise_tax_detail)
-		for key, value in item_wise_tax_detail.items():
-			if key == item_code:
-				tax_amount = value[1]
-				break
-
-	return tax_amount
+	doc = frappe.qb.DocType(doctype)
+	stac = frappe.qb.DocType('Sales Taxes and Charges')
+	invoices = frappe.qb.from_(doc).inner_join(stac).on(stac.parent == doc.name)
+	invoices = invoices.select(
+		stac.account_head,
+		stac.base_tax_amount,
+		stac.item_wise_tax_detail,
+		doc.is_return
+	)
+	invoices = invoices.where(doc.docstatus == 1).where(doc.company == filters.get("company"))
+	invoices = invoices.where(doc.posting_date >= filters.get("from_date"))
+	invoices = invoices.where(doc.posting_date <= filters.get("to_date")).run(as_dict=True)
+
+	for inv in invoices:
+		acc = inv["account_head"]
+		if tax_details.get(acc):
+			tax_details[acc]["total_tax"] += inv["base_tax_amount"]
+			# 'item_wise_tax_detail': '{"Item Code":[Tax Rate, Tax Amount]}'
+			item_wise = json.loads(inv["item_wise_tax_detail"])
+			if not inv["is_return"]:
+				for x in item_wise:
+					if item_wise[x][1] > 0.0:
+						tax_details[acc]["taxable_amount"] += item_wise[x][1]*100/item_wise[x][0]
```

Comment:
Can't we just use net total as taxable value? This back calculating may go wrong in many cases. In cases when users may add taxes as actual amount rather than based on net total

Also, my general concern is there might be users who will be using this item tax template config and might not have  the nil rated and exempted checks in Item. How do they come to know of this change. Their reports will go wrong.

The item tax template in KSA Vat Settings kind of becomes redundant now and its not making it independent of item tax template

Reply:
> Can't we just use the net total as taxable value?

In an invoice with 2 items where one is taxed at 5% and the other is zero/exempt, the net total is not the taxable amount. Hence, it won't be possible to consider the net total as total taxable amount.

> This back calculating may go wrong in many cases. In cases when users may add taxes as actual amount rather than based on net total

Yes, agreed. Maybe a better way would be to use the tax rate(set in the account master) to calculate the taxable amount. That way, even if the user add tax as actual amount, we still we be able to calculate the taxable amount. 

But the tax rate field in accounts doctype isn't mandatory. How about replacing Item Tax Template with a tax rate field in the KSA VAT Settings? 

> Also, my general concern is there might be users who will be using this item tax template config and might not have the nil-rated and exempted checks updated in the Item master. How do they come to know of this change? Their reports will go wrong.

We can run a patch to tick `Is Zero Rated/Is Exempt` if that item has the corresponding Item Tax Template in the taxes table in Item master. We can also run a patch to tick those fields in transactions that are already completed.

We could also place a hook on Sales and Purchase Invoices to tick `Is Zero Rated/Is Exempt` (in the transaction as well as in the Item master) if Item Tax Template set is Zero or Exempt.

We could keep this hook for a period of time, update the docs, inform the users and then slowly remove the hook once we're confident enough.

Depending upon Item Tax Template isn't convenient as:
- It isn't a mandatory field in the transaction and people might forget to fill it.
- In some cases, people will add the correct Item Tax Template but will tamper with the taxes table. Item Tax Template remains unaffected by these tampering. In such cases, tax account ledger will have one data and VAT report another, causing audit issues.
- With issues like #29122, people prefer to stick on to Sales/Purchase Taxes and Charges Template

Therefore, I believe, it is always better and convenient to use the data from the taxes table to generate the report.
@deepeshgarg007

Reply:
@deepeshgarg007